dataflow/gpu-workers/e2e_test.py

```diff
@@ -0,0 +1,120 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import subprocess
+import uuid
+
+from google.cloud import storage
+import pytest
+
+SUFFIX = uuid.uuid4().hex[0:6]
+PROJECT = os.environ["GOOGLE_CLOUD_PROJECT"]
+BUCKET_NAME = f"dataflow-gpu-test-{SUFFIX}"
+IMAGE_NAME = f"gcr.io/{PROJECT}/dataflow/gpu-workers/test-{SUFFIX}"
+
+
+@pytest.fixture
+def bucket_name() -> str:
+    client = storage.Client()
+    bucket = client.create_bucket(BUCKET_NAME)
+
+    yield BUCKET_NAME
+
+    bucket.delete()
+
+
+@pytest.fixture
+def image_name() -> str:
+    subprocess.run(
+        [
+            "gcloud",
+            "builds",
+            "submit",
+            f"--project={PROJECT}",
+            f"--tag={IMAGE_NAME}",
+            "--quiet",
+        ],
+        check=True,
+    )
+
+    yield IMAGE_NAME
+
+    subprocess.run(
+        [
+            "gcloud",
+            "container",
+            "images",
+            "delete",
+            IMAGE_NAME,
+            f"--project={PROJECT}",
+            "--quiet",
+        ],
+        check=True,
+    )
+
+
+def test_python_version() -> None:
+    # Make sure the local and Docker Python versions are the same.
+    # If this test fails, the following needs updating:
+    # - Dockerfile: The `COPY --from=apache/beam` for the worker boot file.
+    # - Docs tutorial: https://cloud.google.com/dataflow/docs/samples/satellite-images-gpus
+    pass
+
+
+def test_apache_beam_version() -> None:
+    # Make sure the installed Apache Beam version matches the Apache Beam image
+    # we use to copy the worker boot file.
+    # If this test fails, the following needs updating:
+    # - Dockerfile: The `COPY --from=apache/beam` for the worker boot file.
+    pass
+
+
+def test_tensorflow_version() -> None:
+    # Make sure the installed Tensorflow version matches the Tensorflow version
+    # in the Dockerfile.
+    # If this test fails, the following needs updating:
+    # - Dockerfile: The `FROM tensorflow/tensorflow` version.
+    pass
+
+
+def test_end_to_end(bucket_name: str, image_name: str) -> None:
+    # Run the Beam pipeline in Dataflow.
+    subprocess.run(
+        [
+            "python",
+            "landsat_view.py",
+            "--scene=LC08_L1TP_115078_20200608_20200625_01_T1",
+            f"--output-path-prefix=gs://{bucket_name}/outputs/",
+            "--runner=DataflowRunner",
+            f"--project={PROJECT}",
+            "--region=us-central1",
+            "--worker_machine_type=custom-1-13312-ext",
+            f"--worker_harness_container_image={image_name}",
+            "--worker_zone=us-central1-a",
+            # "--experiments=worker_accelerator=type=nvidia-tesla-t4,count=1,install-nvidia-driver",
+            "--experiments=use_runner_v2",
+        ],
+        check=True,
+    )
+
+    client = storage.Client()
+    output_size = (
+        client.get_bucket(bucket_name)
```

Comment:
is bucket name regenerated for each test run? otherwise test can reuse the bucket from prior run and interpret stale data as a successful run.

Reply:
A new unique bucket is created, used, and deleted for each test run. This is to avoid concurrent test runs to overlap with each other and avoid race conditions.